File: WordPress/Classes/Services/SuggestionService.m
#import "SuggestionService.h"
#import "Suggestion.h"
#import "AccountService.h"
#import "ContextManager.h"
#import "WPAccount.h"
#import "BlogService.h"
#import "Blog.h"
#import "WordPress-Swift.h"

NSString * const SuggestionListUpdatedNotification = @"SuggestionListUpdatedNotification";

@interface SuggestionService ()

@property (nonatomic, strong) NSMutableArray *siteIDsCurrentlyBeingRequested;

@end

@implementation SuggestionService

+ (instancetype)sharedInstance
{
    static SuggestionService *shared = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        shared = [[self alloc] init];
    });
    return shared;
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        _siteIDsCurrentlyBeingRequested = [NSMutableArray new];
    }
    return self;
}

#pragma mark -

- (NSArray <UserAutocomplete *>*)suggestionsForSiteID:(NSNumber *)siteID
{
    Autocompleter *autocompleter = [self retrieveAutocompleterForSiteID:siteID];

    if (!autocompleter.userAutocompletes) {
        [self updateSuggestionsForSiteID:siteID];
    }
    return [autocompleter.userAutocompletes allObjects];
}

- (void)updateSuggestionsForSiteID:(NSNumber *)siteID
{
    // if there is already a request in place for this siteID, just wait
    if ([self.siteIDsCurrentlyBeingRequested containsObject:siteID]) {
        return;
    }
    
    // add this siteID to currently being requested list
    [self.siteIDsCurrentlyBeingRequested addObject:siteID];
    
    NSString *suggestPath = @"rest/v1.1/users/suggest";
    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
    AccountService *accountService = [[AccountService alloc] initWithManagedObjectContext:context];
    WPAccount *defaultAccount = [accountService defaultWordPressComAccount];
    NSDictionary *params = @{@"site_id": siteID};
    
    __weak __typeof(self) weakSelf = self;
    
    [[defaultAccount wordPressComRestApi] GET:suggestPath
                                   parameters:params
                                      success:^(id responseObject, NSHTTPURLResponse *httpResponse) {

        NSArray *restSuggestions = responseObject[@"suggestions"];

        [self updateUserAutocompletesForSiteID:siteID suggestions:restSuggestions];

        // remove siteID from the currently being requested list
        [weakSelf.siteIDsCurrentlyBeingRequested removeObject:siteID];
    } failure:^(NSError *error, NSHTTPURLResponse *httpResponse){
        // remove siteID from the currently being requested list
        [weakSelf.siteIDsCurrentlyBeingRequested removeObject:siteID];

        DDLogVerbose(@"[Rest API] ! %@", [error localizedDescription]);
    }];
}

- (Autocompleter *)retrieveAutocompleterForSiteID:(NSNumber *)siteID
{
    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];

    NSFetchRequest *request = Autocompleter.fetchRequest;
    [request setPredicate:[NSPredicate predicateWithFormat:@"siteID == %@", siteID]];
    NSError *error;
    NSArray *sites = [context executeFetchRequest:request error:&error];
    if (error) {
        DDLogError(@"Could not fetch Autocompleter for site %@, error", siteID, error);
        return nil;
    } else if (sites.count > 1) {
        DDLogError(@"Retrieved more than one Autocompleter for site %@", siteID);
        return nil;
    } else {
        return [sites firstObject];
    }
}

- (Autocompleter *)insertAutocompleterForSiteID:(NSNumber *)siteID
{
    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
    Autocompleter *autocompleter = [NSEntityDescription insertNewObjectForEntityForName:@"Autocompleter" inManagedObjectContext:context];
    autocompleter.siteID = siteID;
    return autocompleter;
}

- (void)deleteUserAutocompletesForAutocompleter:(Autocompleter *)autocompleter
{
    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
    for (UserAutocomplete *userAutocomplete in autocompleter.userAutocompletes) {
        [context deleteObject:[context objectWithID:userAutocomplete.objectID]];
    }
}

- (void)updateUserAutocompletesForSiteID:(NSNumber *)siteID suggestions:(NSArray *)suggestions
{
    Autocompleter *autocompleter = [self retrieveAutocompleterForSiteID:siteID];

    if (autocompleter) {
        [self deleteUserAutocompletesForAutocompleter:autocompleter];
    } else {
        autocompleter = [self insertAutocompleterForSiteID:siteID];
    }

    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
    for (NSDictionary *suggestion in suggestions) {
        UserAutocomplete *userAutocomplete = [NSEntityDescription insertNewObjectForEntityForName:@"UserAutocomplete" inManagedObjectContext:context];
        userAutocomplete.username = [suggestion stringForKey:@"user_login"];
        userAutocomplete.displayName = [suggestion stringForKey:@"display_name"];
        userAutocomplete.imageURL = [NSURL URLWithString:[suggestion stringForKey:@"image_URL"]];
        userAutocomplete.autocompleter = autocompleter;
    }

    NSError *error;
    if (![context save:&error]) {
        DDLogError(@"Can't save UserAutocompletes for site %@, error: %@", siteID, error);
    } else {
        // send the siteID with the notification so it could be filtered out
        [[NSNotificationCenter defaultCenter] postNotificationName:SuggestionListUpdatedNotification object:siteID];
    }
}

- (BOOL)shouldShowSuggestionsForSiteID:(NSNumber *)siteID
{
    if (!siteID) {
        return NO;
    }

    // if the device is offline
    if (![WordPressAppDelegate shared].connectionAvailable) {
        return NO;
    }
    
    NSManagedObjectContext *context = [[ContextManager sharedInstance] mainContext];
    NSFetchRequest *request = UserAutocomplete.fetchRequest;
    NSError *error;
    NSUInteger count = [context countForFetchRequest:request error:&error];
        
    // if there is nothing to show
    if (count == 0) {
        return NO;
    }
    
    // if the site is not hosted on WordPress.com
    BlogService *service            = [[BlogService alloc] initWithManagedObjectContext:context];
    Blog *blog                      = [service blogByBlogId:siteID];
    return [blog supports:BlogFeatureMentions];
}

@end

File: WordPress/Classes/Services/SuggestionService.swift
import Foundation

/// A service to fetch and persist a list of users that can be @-mentioned in a post or comment.
class SuggestionService {

    private var blogsCurrentlyBeingRequested = [NSNumber]()
    private var requests = [NSNumber: Date]()

    static let shared = SuggestionService()

    /**
    Fetch cached suggestions if available, otherwise from the network if the device is online.

    @param the blog/site to retrieve suggestions for
    @param completion callback containing list of suggestions, or nil if unavailable
    */
    func suggestions(for blog: Blog, completion: @escaping ([UserSuggestion]?) -> Void) {

        let throttleDuration: TimeInterval = 60 // seconds
        let isBelowThrottleThreshold: Bool
        if let id = blog.dotComID, let requestDate = requests[id] {
            isBelowThrottleThreshold = Date().timeIntervalSince(requestDate) < throttleDuration
        } else {
            isBelowThrottleThreshold = false
        }

        if isBelowThrottleThreshold, let suggestions = retrievePersistedSuggestions(for: blog), suggestions.isEmpty == false {
            completion(suggestions)
        } else if ReachabilityUtils.isInternetReachable() {
            fetchAndPersistSuggestions(for: blog, completion: completion)
        } else {
            completion(nil)
        }
    }

    /**
    Performs a REST API request for the given blog.
    Persists response objects to Core Data.

    @param blog/site to retrieve suggestions for
    */
    private func fetchAndPersistSuggestions(for blog: Blog, completion: @escaping ([UserSuggestion]?) -> Void) {

        guard let blogId = blog.dotComID else { return }

        // if there is already a request in place for this blog, just wait
        guard !blogsCurrentlyBeingRequested.contains(blogId) else { return }

        guard let siteID = blog.dotComID else { return }

        let suggestPath = "rest/v1.1/users/suggest"
        let params = ["site_id": siteID]

        // add this blog to currently being requested list
        blogsCurrentlyBeingRequested.append(blogId)

        defaultAccount()?.wordPressComRestApi.GET(suggestPath, parameters: params, success: { [weak self] responseObject, httpResponse in
            guard let `self` = self else { return }
            guard let payload = responseObject as? [String: Any] else { return }
            guard let restSuggestions = payload["suggestions"] as? [[String: Any]] else { return }

            let context = ContextManager.shared.mainContext

            // Delete any existing `UserSuggestion` objects
            self.retrievePersistedSuggestions(for: blog)?.forEach { suggestion in
                context.delete(suggestion)
            }

            // Create new `UserSuggestion` objects
            let suggestions = restSuggestions.compactMap { UserSuggestion(dictionary: $0, context: context) }

            // Associate `UserSuggestion` objects with blog
            blog.userSuggestions = Set(suggestions)

            // Save the changes
            try? blog.managedObjectContext?.save()

            self.requests[blogId] = Date()

            completion(suggestions)

            // remove blog from the currently being requested list
            self.blogsCurrentlyBeingRequested.removeAll { $0 == blogId }
        }, failure: { [weak self] error, _ in
            guard let `self` = self else { return }

            completion(nil)

            // remove blog from the currently being requested list
            self.blogsCurrentlyBeingRequested.removeAll { $0 == blogId}

            DDLogVerbose("[Rest API] ! \(error.localizedDescription)")
        })
    }

    /**
    Tells the caller if it is a good idea to show suggestions right now for a given blog/site.

    @param blog blog/site to check for
    @return BOOL Whether the caller should show suggestions
    */
    func shouldShowSuggestions(for blog: Blog) -> Bool {

        // The device must be online or there must be already persisted suggestions
        guard ReachabilityUtils.isInternetReachable() || retrievePersistedSuggestions(for: blog)?.isEmpty == false else {
            return false
        }

        return blog.supports(.mentions)
    }

    private func defaultAccount() -> WPAccount? {
        let context = ContextManager.shared.mainContext
        let accountService = AccountService(managedObjectContext: context)
        return accountService.defaultWordPressComAccount()
    }

    private func retrievePersistedSuggestions(for blog: Blog) -> [UserSuggestion]? {
        guard let suggestions = blog.userSuggestions else { return nil }
        return Array(suggestions)
    }
}
